tests/unit/test_symmetric_crypto.cpp

```diff
@@ -0,0 +1,312 @@
+{
+    void SetUp() override  {
+        SymmetricCipherBase::SetUp();
+    }
+};
+
+
+TEST_P(SymmetricCipherTest, EncryptDecryptDifferentPlaintextLength)
+{
+    auto operationMode = std::get<0>(GetParam());
+    auto plaintextString = std::get<1>(GetParam());
+    std::vector<uint8_t> plaintext{plaintextString.begin(), plaintextString.end()};
+
+    // When: encrypt and then decrypt the plaintext
+    auto encryptoBuilder = AESCipherBuilder{operationMode, _secretKey};
+    std::shared_ptr<SymmetricCipher> encryptor;
+    if (isAuthenticatedCipherMode(operationMode)) {
+        encryptor = encryptoBuilder.buildAuthenticatedEncryptor();
+    }
+    else {
+        encryptor = encryptoBuilder.buildEncryptor();
+    }
+
+    encryptor->update(plaintext);
+    auto ciphertext = encryptor->finish();
+    auto iv = encryptor->getIV();
+
+    std::vector<uint8_t> tag;
+    auto authenticatedEncryptor = dynamic_cast<AuthenticatedEncryptionI*>(encryptor.get());
+    if (authenticatedEncryptor) {
+        tag = authenticatedEncryptor->getAuthTag();
+    }
+
+    auto decryptorBuilder = AESCipherBuilder{operationMode, _secretKey}.setIV(iv);
+    std::shared_ptr<SymmetricCipher> decryptor;
+    if (isAuthenticatedCipherMode(operationMode)) {
+        decryptor = encryptoBuilder.buildAuthenticatedDecryptor();
+    }
+    else {
+        decryptor = encryptoBuilder.buildDecryptor();
+    }
+
+    auto authenticatedDecryptor = dynamic_cast<AuthenticatedEncryptionI*>(decryptor.get());
+    if (authenticatedDecryptor) {
+        authenticatedDecryptor->setAuthTag(tag);
+    }
+
+    decryptor->update(ciphertext);
+    auto decryptedText = decryptor->finish();
+
+    // Then: Decrypted text should match the plaintext
+    ASSERT_THAT(decryptedText, ::testing::ElementsAreArray(plaintext));
+}
+
+
+static std::vector<EncrytDecryptTestData> prepareTestDataForMode(SymmetricCipherMode mode)
+{
+    static const std::vector<std::string> PlaintextStringsOfDifferentLength
+            {"0",
+             "0123456789", // less than a block
+             "0123456789123456", // a block
+             "01234567891234567890" // more than a block
+            };
+
+    std::vector<EncrytDecryptTestData> testData;
+    for (const auto &elem : PlaintextStringsOfDifferentLength) {
+        testData.emplace_back(std::make_tuple(mode, elem));
+    }
+    return testData;
+}
+
+INSTANTIATE_TEST_SUITE_P(GCM,
+                         SymmetricCipherTest,
+                         testing::ValuesIn(prepareTestDataForMode(SymmetricCipherMode::GCM)));
+
+INSTANTIATE_TEST_SUITE_P(CBC,
+                         SymmetricCipherTest,
+                         testing::ValuesIn(prepareTestDataForMode(SymmetricCipherMode::CBC)));
+
+INSTANTIATE_TEST_SUITE_P(CTR,
+                         SymmetricCipherTest,
+                         testing::ValuesIn(prepareTestDataForMode(SymmetricCipherMode::CTR)));
+
+
+TEST_P(SymmetricCipherAdvancedTest, encryptMultipleChunksAndDecrypt)
```

Comment:
Should we add tests to ensure that decryption fails with wrong key/iv/ciphertext and/or decryption with AEAD ciphers fails when the ciphertext was modified?

Reply:
The patch is already big enough. Shouldn't we add extra tests in a separate MR?